src/ReferenceList.php

```diff
@@ -70,6 +70,17 @@ public function addReference( Reference $reference, $index = null ) {
 		}
 	}
 
+	/**
+	 * @see parent::attach
+	 * @param Reference $reference
+	 * @param mixed $data
+	 */
+	public function attach( $reference, $data = null ) {
+		if ( !$reference->isEmpty() ) {
```

Comment:
I think you need to add this check to `addReference`. Calling `addReference( new Reference(), 0 )` will still be allowed but shouldn't for consistency.